ToolBox/PsychSignalDataConverter/PsychSignalDataDownloader.cs

```diff
@@ -0,0 +1,149 @@
+            }
+            
+            // PsychSignal paginates data by hour
+            for (; startDateUtc < endDateUtc; startDateUtc = startDateUtc.AddHours(1))
+            {
+                var rawDataPath = Path.Combine(_rawDataDestination, $"{startDateUtc:yyyyMMdd_HH}_{_dataSource}.csv");
+                var rawDataPathTemp = Path.Combine(Path.GetTempPath(), $"{startDateUtc:yyyyMMdd_HH}_{_dataSource}.csv.tmp");
+                
+                // Don't download files we already have
+                if (File.Exists(rawDataPath))
+                {
+                    continue;
+                }
+                
+                // Retry in case a download failed
+                for (var retries = 0; retries < MaxRetries; retries++)
+                {
+                    // Psychsignal imposes very strict rate limits
+                    Thread.Sleep(10000);
+
+                    try
+                    {
+                        using (var client = new WebClient())
+                        {
+                            client.DownloadFile($"{_baseUrl}/replay?apikey={_apiKey}&update=1m&sources={_dataSource}&from={startDateUtc:yyyyMMddHH}&format=csv", rawDataPathTemp);
+                            File.Move(rawDataPathTemp, rawDataPath);
+                            Log.Trace($"PsychSignalDataDownloader.Download(): Successfully downloaded file: {rawDataPath}");
+                            break;
+                        }
+                    }
+                    catch (WebException e)
+                    {
+                        var response = (HttpWebResponse) e.Response;
+
+                        if (retries == MaxRetries - 1)
+                        {
+                            Log.Error($"PsychSignalDataDownloader.Download(): We've reached the maximum number of retries for date {startDateUtc:yyyy-MM-dd HH:00:00}");
+                            continue;
+                        }
+                        if (response == null)
+                        {
+                            Log.Error("PsychSignalDataDownloader.Download(): Response was null. Retrying...");
+                            continue;
+                        }
+                        if (response.StatusCode == HttpStatusCode.BadRequest)
+                        {
+                            Log.Error("PsychSignalDataDownloader.Download(): Server received a bad request. Continuing...");
+                            break;
+                        }
+                        if (response.StatusCode == HttpStatusCode.NotFound)
+                        {
+                            Log.Error("PsychSignalDataDownloader.Download(): Received an HTTP 404. Continuing...");
+                            break;
+                        }
+                        if (response.StatusCode == (HttpStatusCode) 429)
+                        {
+                            Log.Trace("PsychSignalDataDownloader.Download(): We are being rate limited. Retrying...");
```

Comment:
Would the sleep be better placed here instead of at the top of the loop? Also, for documentation purposes, it may be helpful to include the API's rat limiting rules.

Reply:
Every request, no matter if it's successful or not, requires that it be sent `n` seconds after the previous request, in which this case I've set `n` to 10 seconds. 

They don't document their API. I had to figure out an appropriate rate limit myself through trial and error

Reply:
oh wow, don't you just _love_ undocumented APIs! 👍 

Still thinking a `RateGate` would be more explicit as to the allowable rate limits (empirically set at 1 call per 10 seconds ?)